pkg/scheduler/queue/query_component_capacity.go

```diff
@@ -0,0 +1,177 @@
+	// Annotations from the frontend representing "both" or "unknown" do not need to be checked explicitly.
+	// Conservatively, we assume a query request will be served by both query components
+	// as we prefer to overestimate query component capacity utilization rather than underestimate.
+	isIngester, isStoreGateway = true, true
+
+	switch queryComponentName {
+	case ingesterQueueDimension:
+		isStoreGateway = false
+	case storeGatewayQueueDimension:
+		isIngester = false
+	}
+	return isIngester, isStoreGateway
+}
+
+// QueryComponentCapacity tracks requests from the time they are forwarded to a querier
+// to the time are completed by the querier or failed due to cancel, timeout, or disconnect.
+// Unlike the Scheduler's schedulerInflightRequests, tracking begins only when the request is sent to a querier.
+//
+// Scheduler-Querier inflight requests are broken out by the query component flags,
+// representing whether the query request will be served by the ingesters, store-gateways, or both.
+// Query requests utilizing both ingesters and store-gateways are tracked in the atomics for both component,
+// therefore the sum of inflight requests by component is likely to exceed the inflight requests total.
+type QueryComponentCapacity struct {
+	// targetReservedCapacity sets the portion of querier-worker connections we attempt to reserve
+	// for queries to the less-loaded query component when the query queue becomes backlogged.
+	targetReservedCapacity float64
+
+	ingesterInflightRequests     *atomic.Int64
+	storeGatewayInflightRequests *atomic.Int64
+	querierInflightRequestsTotal *atomic.Int64
+
+	querierInflightRequestsGauge *prometheus.GaugeVec
+}
+
+// DefaultReservedQueryComponentCapacity reserves 1 / 3 of querier-worker connections
+// for the query component utilizing fewer of the available connections.
+const DefaultReservedQueryComponentCapacity = 0.33
+
+const MinReservedQueryComponentCapacity = 0.1
+const MaxReservedQueryComponentCapacity = 0.5
+
+func NewQueryComponentCapacity(
+	targetReservedCapacity float64,
+	querierInflightRequests *prometheus.GaugeVec,
+) (*QueryComponentCapacity, error) {
+
+	if targetReservedCapacity < MinReservedQueryComponentCapacity || targetReservedCapacity >= MaxReservedQueryComponentCapacity {
+		return nil, errors.New("invalid targetReservedCapacity")
+	}
+
+	return &QueryComponentCapacity{
+		targetReservedCapacity: targetReservedCapacity,
+
+		ingesterInflightRequests:     atomic.NewInt64(0),
+		storeGatewayInflightRequests: atomic.NewInt64(0),
+		querierInflightRequestsTotal: atomic.NewInt64(0),
+
+		querierInflightRequestsGauge: querierInflightRequests,
+	}, nil
+}
+
+// ExceedsCapacityForComponentName checks whether a query component has exceeded the capacity utilization threshold.
+// This enables the dequeuing algorithm to skip requests for a query component experiencing heavy load,
+// reserving querier-worker connection capacity to continue servicing requests for the other component.
```

Comment:
I know you haven't implemented it yet, but this isn't quite how I imagined the dequeuing working.

tl;dr: querier workers should never be idle if there are requests in the queue, but this comment makes it sound like we would keep some workers idle

Rather than skipping requests for a query component experiencing heavy load and reserving capacity for requests for the other component, I think we should instead just prioritise requests for the component under low load. This way, if (say) store-gateways are under high load, but there are no waiting requests for ingesters, we'll still make use of the available querier capacity to work on store-gateway requests. But, if ingester requests later arrive, we'd then pick those to work on next over other older requests for store-gateways.